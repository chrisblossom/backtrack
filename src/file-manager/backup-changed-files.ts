import path from 'path';
import { rootPath } from '../config/paths';
import { backupFile } from '../utils/backup-file';
import log from '../utils/log';

import { ParsedFiles, FileStats } from '../types';

async function backupChangedFiles(
	files: ParsedFiles,
	previousStats: FileStats = {},
) {
	const backupNewFiles = files.dest.files
		.map((destFile) => {
			/**
			 * This only handles new files
			 */
			const previousFileHash = previousStats[destFile];
			if (previousFileHash) {
				return null;
			}

			const destFileIndex = files.dest.files.indexOf(destFile);
			const srcFile = files.src.files[destFileIndex];
			const srcHash = files.src.hash[srcFile];
			const destHash = files.dest.hash[destFile];
			const allowChanges = files.dest.allowChanges[destFile];
			const destExists = !!destHash;

			/**
			 * Don't backup if file doesn't exist or the files are the same
			 */
			if (!destExists || srcHash === destHash) {
				return null;
			}

			/**
			 * Do not backup file if changes are allowed
			 */
			if (allowChanges === true) {
				return null;
			}

			const originalFilePath = files.dest.absolute[destFile];
			return backupFile(originalFilePath).then((backupResult) => {
				if (backupResult) {
					const relativeBackupFilePath = path.relative(
						rootPath,
						backupResult.file,
					);

					const relativeFilePath = path.relative(
						rootPath,
						originalFilePath,
					);

					log.warn(
						`${relativeFilePath} already exists. Moving to ${relativeBackupFilePath}`,
					);
				}

				return backupResult;
			});
		})
		.filter(Boolean);

	const backupOldFiles = Object.keys(previousStats)
		.map((destFile) => {
			const destHash = files.dest.hash[destFile];

			const previousFileHash = previousStats[destFile];
			const destExists = !!destHash;

			const destFileIndex = files.dest.files.indexOf(destFile);

			/**
			 * Need to check for undefined because previousStats source file could have been removed
			 */
			const srcFile =
				destFileIndex !== -1
					? files.src.files[destFileIndex]
					: undefined;

			const srcHash = srcFile ? files.src.hash[srcFile] : undefined;

			/**
			 * Do not backup file if changes are allowed
			 */
			const allowChanges = files.dest.allowChanges[destFile];
			if (allowChanges === true) {
				return null;
			}

			/**
			 * Don't backup if dest hasn't changed
			 */
			if (
				destExists === false ||
				previousFileHash === destHash ||
				// already correct hash
				srcHash === destHash
			) {
				return null;
			}

			/**
			 * Move current destFile if modified
			 */
			const originalFilePath = files.dest.absolute[destFile];

			return backupFile(originalFilePath).then((backupResult) => {
				if (backupResult) {
					const relativeBackupFilePath = path.relative(
						rootPath,
						backupResult.file,
					);

					const relativeFilePath = path.relative(
						rootPath,
						originalFilePath,
					);

					log.warn(
						`${relativeFilePath} hash mismatch. Moving to ${relativeBackupFilePath}`,
					);
				}

				return backupResult;
			});
		})
		.filter(Boolean);

	// @ts-ignore
	await Promise.all([...backupNewFiles, ...backupOldFiles]);
}

export { backupChangedFiles };
