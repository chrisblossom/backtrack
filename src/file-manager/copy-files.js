/* @flow */

import path from 'path';
import { copy } from '../utils/copy';
import log from '../utils/log';
import { rootPath } from '../config/paths';

import type { ParsedFiles, FileStats } from '../types.js';

async function copyFiles(files: ParsedFiles, previousStats?: FileStats = {}) {
    const src = files.src;
    const dest = files.dest;

    const filesToCopy = src.files.reduce((acc, srcFile, index) => {
        const destFile = dest.files[index];

        const srcHash = src.hash[srcFile];
        const destHash = dest.hash[destFile];

        /**
         * Only copy files that have changed / do not exist
         *
         * hash === '' means file does not exist
         */
        if (srcHash !== destHash) {
            const srcAbsolute = src.absolute[srcFile];

            if (destHash) {
                const previousHash = previousStats[destFile];
                /**
                 * Do not copy if changes are allowed and destFile has been modified
                 */
                const allowChanges = dest.allowChanges[destFile];
                if (allowChanges === true && destHash !== previousHash) {
                    const destAbsolute = dest.absolute[destFile];
                    const { base: filename, dir, ext = '' } = path.parse(
                        destAbsolute,
                    );

                    const updatedFilename = `${filename}-latest${ext}`;

                    const updatedFileAbsolute = path.resolve(
                        dir,
                        updatedFilename,
                    );

                    const relativeFilePath = path.relative(
                        rootPath,
                        updatedFileAbsolute,
                    );

                    /**
                     * Since backtrack-stats tracks the latest srcHash,
                     * only copy latest file if source has changed.
                     */
                    if (srcHash === previousHash) {
                        return acc;
                    }

                    log.info(
                        `Changed file ${destFile} source updated. Copying to: ${relativeFilePath}`,
                    );

                    const copyFile = {
                        src: srcAbsolute,
                        dest: updatedFileAbsolute,
                    };

                    return [...acc, copyFile];
                }

                log.info(`Updating file: ${destFile}`);
            } else {
                log.info(`Adding file: ${destFile}`);
            }

            const destAbsolute = dest.absolute[destFile];
            const copyFile = {
                src: srcAbsolute,
                dest: destAbsolute,
            };

            return [...acc, copyFile];
        }

        return acc;
    }, []);

    /**
     * Bail early if nothing else to do
     */
    if (filesToCopy.length === 0) {
        return;
    }

    /**
     * Copy all files
     *
     * Must overwrite because source files can change
     * Safe to overwrite because changed dest files have been backed up
     */
    await copy(filesToCopy, {
        overwrite: true,
        errorOnExist: true,
    });
}

export { copyFiles };
