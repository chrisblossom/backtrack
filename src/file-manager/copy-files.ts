import path from 'path';
import { copy, File } from '../utils/copy';
import log from '../utils/log';
import { rootPath } from '../config/paths';
import { ParsedFiles, FileStats } from '../types';

// prettier-ignore
type CopyInfo =
	{ shouldCopy: false } |
	{ shouldCopy: true; destPath: string };

/*
 * Handles the logic for when file changes are allowed.
 * It determines if a file should be copied and creates a new filename with '-latest'.
 */
function handleAllowedChanges(
	files: ParsedFiles,
	destFile: string,
	srcHash: string,
	previousHash: string,
): CopyInfo {
	const destAbsolute = files.dest.absolute[destFile];
	const { base: filename, dir, ext = '' } = path.parse(destAbsolute);
	const updatedFilename = `${filename}-latest${ext}`;
	const updatedFileAbsolute = path.resolve(dir, updatedFilename);
	const relativeFilePath = path.relative(rootPath, updatedFileAbsolute);

	/**
	 * Since backtrack-stats tracks the latest srcHash,
	 * only copy latest file if source has changed.
	 */
	if (srcHash === previousHash) {
		return { shouldCopy: false };
	}

	log.info(
		`Changed file ${destFile} source updated. Copying to: ${relativeFilePath}`,
	);

	return { shouldCopy: true, destPath: updatedFileAbsolute };
}

/*
 * Determines whether a file should be copied based on its hash and other criteria.
 */
function shouldCopyFile(
	files: ParsedFiles,
	destFile: string,
	srcHash: string,
	destHash: string,
	previousStats: FileStats,
): CopyInfo {
	if (destHash) {
		const previousHash = previousStats[destFile];
		const ignoreUpdates = files.dest.ignoreUpdates[destFile];

		/**
		 * Do not copy if ignoreUpdates is true and file already exists
		 */
		if (ignoreUpdates && previousHash !== srcHash) {
			log.info(
				`Unmanaged file ${destFile} source changed. Updating .backtrack-stats.json`,
			);
			return { shouldCopy: false };
		}

		/**
		 * Do not copy if changes are allowed and destFile has been modified
		 */
		const allowChanges = files.dest.allowChanges[destFile];
		if (allowChanges && destHash !== previousHash) {
			return handleAllowedChanges(files, destFile, srcHash, previousHash);
		}

		log.info(`Updating file: ${destFile}`);
	} else {
		log.info(`Adding file: ${destFile}`);
	}

	return { shouldCopy: true, destPath: files.dest.absolute[destFile] };
}

/*
 * Creates a list of files to be copied by evaluating each source file.
 */
function getFilesToCopy(files: ParsedFiles, previousStats: FileStats): File[] {
	return files.src.files.reduce<File[]>((acc, srcFile, index) => {
		const destFile = files.dest.files[index];
		const srcHash = files.src.hash[srcFile];
		const destHash = files.dest.hash[destFile];

		if (srcHash !== destHash) {
			const srcAbsolute = files.src.absolute[srcFile];
			const destAbsolute = files.dest.absolute[destFile];

			const copyInfo = shouldCopyFile(
				files,
				destFile,
				srcHash,
				destHash,
				previousStats,
			);

			if (copyInfo.shouldCopy) {
				acc.push({
					src: srcAbsolute,
					dest: copyInfo.destPath ?? destAbsolute,
				});
			}
		}

		return acc;
	}, []);
}

/*
 * Main function to copy files.
 * It determines which files need to be copied and performs the copy operation.
 */
async function copyFiles(
	files: ParsedFiles,
	previousStats: FileStats = {},
): Promise<void> {
	const filesToCopy = getFilesToCopy(files, previousStats);

	// nothing to copy
	if (filesToCopy.length === 0) {
		return;
	}

	/**
	 * Copy all files
	 *
	 * Must overwrite because source files can change
	 * Safe to overwrite because changed dest files have been backed up
	 */
	await copy(filesToCopy);
}

export { copyFiles };
