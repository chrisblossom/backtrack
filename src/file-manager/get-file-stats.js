/* @flow */

import type { ParsedFiles, FileManagerStats } from '../types.js';

function getFileStats(parsedFiles: ParsedFiles): FileManagerStats {
    const fileHashes = parsedFiles.dest.files.reduce((acc, destFile) => {
        const destFileIndex = parsedFiles.dest.files.indexOf(destFile);

        /**
         * Use source file's hash to allow for 'allowChanges' option and guard against file system errors
         *
         * destHash will equal srcHash because the copy_files task is done prior to stats
         */
        const srcFile = parsedFiles.src.files[destFileIndex];
        const srcHash = parsedFiles.src.hash[srcFile];

        return {
            ...acc,
            [destFile]: srcHash,
        };
    }, {});

    const stats = {};
    if (Object.keys(fileHashes).length !== 0) {
        stats.files = fileHashes;
    }

    if (parsedFiles.makeDirs.length !== 0) {
        stats.directories = parsedFiles.makeDirs;
    }

    return stats;
}

export { getFileStats };
