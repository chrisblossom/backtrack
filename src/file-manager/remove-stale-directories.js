/* @flow */

import path from 'path';
import { existsSync, readdir } from 'fs';
import del from 'del';
import { rootPath } from '../config/paths';
import log from '../utils/log';
import { normalizePathname } from '../utils/normalize-pathname';

import type { ParsedFiles, DirStats } from '../types.js';

function readDir(pathname) {
    return new Promise((resolve, reject) => {
        readdir(pathname, (error, list) => {
            if (error) {
                reject(error);
                return;
            }

            resolve(list);
        });
    });
}

async function shouldDelete(relativeDir, makeDirs = []) {
    async function goUpOne() {
        const upOneDir = path.join(relativeDir, '..');
        if (upOneDir === '.') {
            return;
        }

        await shouldDelete(upOneDir, makeDirs);
    }

    if (makeDirs.includes(relativeDir) === true) {
        return;
    }

    const absolutePath = path.resolve(rootPath, relativeDir);
    const exists = existsSync(absolutePath);
    if (exists === false) {
        await goUpOne();

        return;
    }

    const directoryList = await readDir(absolutePath);
    if (directoryList.length !== 0) {
        log.info(`Directory not empty: ${relativeDir}`);

        return;
    }

    if (path.resolve(relativeDir) === rootPath) {
        return;
    }

    log.info(`Removing directory: ${relativeDir}`);

    await del(absolutePath);

    await goUpOne();
}

async function removeStaleDirectories(
    parsedFiles: ParsedFiles,
    previousStats?: DirStats = [],
) {
    for (const dir of previousStats) {
        const normalizedDir = normalizePathname(dir);

        /**
         * Run sequentially to ensure directories are empty before moving to next
         */
        // eslint-disable-next-line no-await-in-loop
        await shouldDelete(normalizedDir, parsedFiles.makeDirs);
    }
}

export { removeStaleDirectories };
