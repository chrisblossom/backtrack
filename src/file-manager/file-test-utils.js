/* @flow */

/* eslint-disable arrow-body-style */

import type { FileManager } from '../types.js';

const filesPreprocessor = (args) =>
    // $FlowIgnore
    require('../options-file/files-preprocessor').filesPreprocessor(args);
const filesParser = (args) =>
    // $FlowIgnore
    require('../options-file/files-post-processor').filesPostProcessor(args);
const getFileStats = (parsedFiles) =>
    require('./get-file-stats').getFileStats(parsedFiles);

function fileStats(previousFiles: FileManager) {
    const normalize = filesPreprocessor({ value: previousFiles });

    const parsedPreviousFiles = filesParser({ value: normalize });

    const previousStats = getFileStats(parsedPreviousFiles).files;

    return previousStats;
}

function fileInfo(files: FileManager = [], previousFiles: FileManager = []) {
    const previousStats = fileStats(previousFiles);
    const normalize = filesPreprocessor({ value: files });

    const parsedFiles = filesParser({ value: normalize });

    return {
        previousStats,
        parsedFiles,
    };
}

export { fileInfo, fileStats };
