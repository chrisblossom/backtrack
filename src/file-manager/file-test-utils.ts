/* eslint-disable arrow-body-style,@typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-return,@typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-assignment */

import { FileManager, FileManagerStats } from '../types';

const filesPreprocessor = (args: any) =>
	require('../options-file/files-preprocessor').filesPreprocessor(args);
const filesParser = (args: any) =>
	require('../options-file/files-post-processor').filesPostProcessor(args);
const getFileStats = (parsedFiles: any) =>
	require('./get-file-stats').getFileStats(parsedFiles);

function fileStats(previousFiles: FileManager): FileManagerStats {
	const normalize = filesPreprocessor({ value: previousFiles });

	const parsedPreviousFiles = filesParser({ value: normalize });

	const previousStats = getFileStats(parsedPreviousFiles).files;

	return previousStats;
}

interface FileInfo {
	previousStats: FileManagerStats;
	parsedFiles: FileManager;
}

function fileInfo(
	files: FileManager = [],
	previousFiles: FileManager = [],
): FileInfo {
	const previousStats = fileStats(previousFiles);
	const normalize = filesPreprocessor({ value: files });

	const parsedFiles = filesParser({ value: normalize });

	return {
		previousStats,
		parsedFiles,
	};
}

export { fileInfo, fileStats };
