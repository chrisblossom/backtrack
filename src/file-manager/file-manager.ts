import path from 'path';
import { backupChangedFiles } from './backup-changed-files';
import { removeStaleFiles } from './remove-stale-files';
import { removeStaleDirectories } from './remove-stale-directories';
import { copyFiles } from './copy-files';
import { getFileStats } from './get-file-stats';
import { makeDirs } from '../utils/make-dirs';
import log from '../utils/log';
import { rootPath } from '../config/paths';
import { filesPostProcessor } from '../options-file/files-post-processor';

import { ParsedFiles, FileManagerStats } from '../types';

type Return = Promise<FileManagerStats>;

const base = filesPostProcessor();

async function fileManager(
    files: ParsedFiles = base,
    previousStats: FileManagerStats = {},
): Return {
    /**
     * Create Directories
     */
    const createDirectories = await makeDirs(files.makeDirs);

    createDirectories.forEach((absolutePath) => {
        const relativeDir = path.relative(rootPath, absolutePath);

        log.info(`Creating directory: ${relativeDir}`);
    });

    /**
     * Backup any user-changed files
     */
    await backupChangedFiles(files, previousStats.files);

    /**
     * Remove stale files
     */
    await removeStaleFiles(files, previousStats.files);

    /**
     * Remove stale directories
     */
    await removeStaleDirectories(files, previousStats.directories);

    /**
     * Copy files
     */
    await copyFiles(files, previousStats.files);

    /**
     * Get updated stats
     */
    const fileStats = getFileStats(files);

    return fileStats;
}

export { fileManager };
