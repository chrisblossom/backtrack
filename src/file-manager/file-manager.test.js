/* @flow */

import path from 'path';
import { realpathSync } from 'fs';
import del from 'del';
import os from 'os';
import fse from 'fs-extra';
import { getRandomInteger } from '../utils/get-random-number';
import { fileInfo } from './file-test-utils';

const fileManager = (copyFiles, previousStats) =>
    require('./file-manager').fileManager(copyFiles, previousStats);

describe('fileManger', () => {
    const cwd = process.cwd();
    /**
     * Use os.tmpdir instead of __sandbox__ because Jest cannot ignore dynamic created files in watch mode
     * https://github.com/facebook/jest/issues/3923
     *
     * Use random dir name to ensure unique
     */
    const dir = path.resolve(
        realpathSync(os.tmpdir()),
        'backtrack',
        `cwd_${getRandomInteger()}`,
    );

    beforeEach(async () => {
        jest.mock('../utils/log', () => ({
            warn: jest.fn(),
            info: jest.fn(),
            error: jest.fn(),
            success: jest.fn(),
        }));

        await fse.ensureDir(dir);

        process.chdir(dir);

        await fse.copy(path.resolve(__dirname, '__sandbox__/app1/'), dir);
    });

    afterEach(async () => {
        await del(path.resolve(dir, '**/*'), {
            dot: true,
        });

        process.chdir(cwd);
    });

    afterAll(async () => {
        process.chdir(path.resolve(dir, '../'));

        await del(dir, {
            dot: true,
        });

        process.chdir(cwd);
    });

    test('handles undefined files', async () => {
        const result = await fileManager();

        expect(result).toEqual({});
    });

    test('creates file', async () => {
        const files = require(path.resolve(dir, 'backtrack.config')).files;

        const { parsedFiles } = fileInfo(files);

        const result = await fileManager(parsedFiles);

        expect(result).toMatchSnapshot();
    });
});
