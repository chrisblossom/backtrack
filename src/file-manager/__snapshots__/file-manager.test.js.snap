// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fileManger creates file 1`] = `
Object {
  "files": Object {
    "fake.config.js": "a6a55f434d6dc8a956cd49e7e5e91455deb92f94c5b25d6cdfe678ddfa80792c",
  },
}
`;
